File: 003/Problem3.mjs
import { performance } from "perf_hooks";

function primeFactors(n) {
  var numbers = [];

  // Print the number of 2s that divide n
  while (n % 2 == 0) {
    numbers.push(2);
    n = Math.floor(n / 2);
  }

  // n must be odd at this point.
  // So we can skip one element
  // (Note i = i +2)
  for (let i = 3; i <= Math.floor(Math.sqrt(n)); i = i + 2) {
    // While i divides n, print i
    // and divide n
    while (n % i == 0) {
      numbers.push(i);
      n = Math.floor(n / i);
    }
  }

  // This condition is to handle the
  // case when n is a prime number
  // greater than 2
  if (n > 2) numbers.push(n);

  return numbers[numbers.length - 1];
}

function problem() {
  console.log(primeFactors(600851475143));
}

var start = performance.now();
problem();
var end = performance.now();
console.log("Function took " + (end - start) / 1000);

File: 003/Problem3.go
package main

import (
	"fmt"
	"time"
)

const problemNumber int = 600851475143

// Get all prime factors of a given number n
func PrimeFactors(n int) (pfs []int) {
	// Get the number of 2s that divide n
	for n%2 == 0 {
		pfs = append(pfs, 2)
		n = n / 2
	}

	// n must be odd at this point. so we can skip one element
	// (note i = i + 2)
	for i := 3; i*i <= n; i = i + 2 {
		// while i divides n, append i and divide n
		for n%i == 0 {
			pfs = append(pfs, i)
			n = n / i
		}
	}

	// This condition is to handle the case when n is a prime number
	// greater than 2
	if n > 2 {
		pfs = append(pfs, n)
	}

	return
}

func problem() {
	nums := PrimeFactors(problemNumber)
	fmt.Println(nums[len(nums)-1])
}

func main() {
	start := time.Now()
	for i := 0; i < 1000; i++ {
		problem()
	}
	elapsed := time.Since(start)
	fmt.Printf("Function took %s", elapsed)
}
